code/modules/projectiles/projectile/trace.dm

```diff
@@ -0,0 +1,76 @@
+//Helper proc to check if you can hit them or not.
+/proc/check_trajectory(atom/target as mob|obj, atom/firer as mob|obj, var/pass_flags=PASSTABLE|PASSGLASS|PASSGRILLE, flags=null)
+	if(!istype(target) || !istype(firer))
+		return 0
+
+	var/obj/item/projectile/test/trace = new /obj/item/projectile/test(get_turf(firer)) //Making the test....
+
+	//Set the flags and pass flags to that of the real projectile...
+	if(!isnull(flags))
+		trace.flags = flags
+	trace.pass_flags = pass_flags
+
+	var/output = trace.launch(target) //Test it!
+	qdel(trace) //No need for it anymore
+	return output //Send it back to the gun!
+
+/obj/item/projectile/proc/check_fire(atom/target as mob, var/mob/living/user as mob)  //Checks if you can hit them or not.
+	check_trajectory(target, user, pass_flags, flags)
+
+//"Tracing" projectile
+/obj/item/projectile/test //Used to see if you can hit them.
+	invisibility = 101 //Nope!  Can't see me!
+	yo = null
+	xo = null
+	var/result = 0 //To pass the message back to the gun.
+
+/obj/item/projectile/test/Bump(atom/A as mob|obj|turf|area)
+	if(A == firer)
+		loc = A.loc
+		return //cannot shoot yourself
+	if(istype(A, /obj/item/projectile))
+		return
+	if(istype(A, /mob/living) || istype(A, /obj/mecha) || istype(A, /obj/vehicle))
+		result = 2 //We hit someone, return 1!
+		return
+	result = 1
+	return
+
+/obj/item/projectile/test/launch(atom/target)
```

Comment:
This either needs to be `proc/launch` or renamed to `launch_projectile`. Probably the latter?